Add `cupy.lib.stride_tricks.sliding_window_view`

negin513 opened the pull request:
## Summary

This pull request introduces the `sliding_window_view` function to CuPy, enhancing its alignment with the NumPy API. The addition of this feature allows users to create a sliding window view of n-dimensional arrays, providing powerful and flexible functionality for array manipulation. Please see #6078. 

## Key Changes

- Implements the `sliding_window_view` function in CuPy, similar to its counterpart in the NumPy API.
- Resolves: Issue #6956 and part of #6078.  Also relevant to [cupy-xarray/9](https://github.com/xarray-contrib/cupy-xarray/issues/9)

## Motivation

The integration of this function enables a more seamless transition for users who switch between or use both NumPy and CuPy in their projects. 

Importantly, this feature is critical for Xarray-Cupy integration (see [cupy-xarray/9](https://github.com/xarray-contrib/cupy-xarray/issues/9)) , allowing users to benefit from a more efficient and enhanced use case experience. 

Contributors: @negin513, @dcherian
Looking forward to your feedback!

Thank you for your time and consideration.

takagi commented:
Would you also add some tests for this function? Thanks!

dcherian commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +38,117 @@ def as_strided(x, shape=None, strides=None):
+                    if np.iterable(window_shape)
+                    else (window_shape,))
+
+    # -- because normalize_axis_tuple does not handle lists:
+    # -- with Xarray it causes error : 
+    # -- TypeError: 'list' object cannot be interpreted as an integer
+    axis = (
+        axis[0]
+        if np.iterable(axis) and len(axis) == 1
+        else axis)
```
I think we should fix this in xarray.

Though perhaps cupy should just vendor the upstream `normalize_axis_tuple` which handles lists.
```
np.core.numeric.normalize_axis_tuple([1, 2, -1], 4) # (1, 2, 3)
```

negin513 replied:
Thanks and done! An issue regarding this is also filed with Xarray :  https://github.com/pydata/xarray/issues/7890

dcherian commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +38,117 @@ def as_strided(x, shape=None, strides=None):
+    positions.
+
+
+	Parameters
```
```suggestion
Parameters
```

dcherian commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +38,117 @@ def as_strided(x, shape=None, strides=None):
+    # -- because normalize_axis_tuple does not handle lists:
+    # -- with Xarray it causes error : 
+    # -- TypeError: 'list' object cannot be interpreted as an integer
+    axis = (
+        axis[0]
+        if np.iterable(axis) and len(axis) == 1
+        else axis)
```
```suggestion
```

negin513 replied:
done! Thanks!

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -29,6 +29,65 @@ def test_rolling_window(self, xp):
 
 
+class TestSlidingWindowView:
+    def test_1d(self):
```
Could you use `testing.numpy_cupy_array_equal()` to compare the outputs of `sliding_window_view` between NumPy and CuPy?

negin513 replied:
`@testing.numpy_cupy_array_equal()` is added for all tests and is passing now...

takagi commented:
/test mini

negin513 commented:
> Would you also add some tests for this function? Thanks!

Tests are now added.

dcherian commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -29,6 +29,38 @@ def test_rolling_window(self, xp):
+        arr_view = sliding_window_view_cp(arr, window_shape=window_shape)
+        return arr_view
+
+    @testing.numpy_cupy_array_equal()
+    def test_2d_with_axis(self):
+        arr = testing.shaped_arange((3,4))
+        window_shape = 3
+        axis = 0
+        arr_view = sliding_window_view_cp(arr, window_shape,axis)
+        return arr_view
+
+    @testing.numpy_cupy_array_equal()
+    def test_2d_multi_axis(self):
+        arr = testing.shaped_arange((3,4))
+        window_shape = (2,3)
+        axis = (0,1)
+        arr_view = sliding_window_view_cp(arr, window_shape,axis)
+        return arr_view
+
```
```suggestion
    @testing.numpy_cupy_array_equal()
    def test_2d_with_axis(self, xp):
        arr = testing.shaped_arange((3,4), xp)
        window_shape = 3
        axis = 0
        arr_view = stride_tricks.sliding_window_view(arr, window_shape,axis)
        return arr_view

    @testing.numpy_cupy_array_equal()
    def test_2d_multi_axis(self):
        arr = testing.shaped_arange((3,4), xp)
        window_shape = (2,3)
        axis = (0,1)
        arr_view = stride_tricks.sliding_window_view(arr, window_shape,axis)
        return arr_view

```

dcherian commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -29,6 +29,38 @@ def test_rolling_window(self, xp):
+        arr_view = sliding_window_view_cp(arr, window_shape=window_shape)
+        return arr_view
+
+    @testing.numpy_cupy_array_equal()
+    def test_2d_with_axis(self):
+        arr = testing.shaped_arange((3,4))
+        window_shape = 3
+        axis = 0
+        arr_view = sliding_window_view_cp(arr, window_shape,axis)
+        return arr_view
```
```suggestion
```suggestion
    @testing.numpy_cupy_array_equal()
    def test_2d_with_axis(self, xp):
        arr = testing.shaped_arange((3,4), xp)
        window_shape = 3
        axis = 0
        arr_view = stride_tricks.sliding_window_view(arr, window_shape,axis)
        return arr_view
```

dcherian commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -29,6 +29,38 @@ def test_rolling_window(self, xp):
+        return arr_view
+
+    @testing.numpy_cupy_array_equal()
+    def test_2d_with_axis(self):
+        arr = testing.shaped_arange((3,4))
```
```suggestion
    def test_2d_with_axis(self, xp):
        arr = testing.shaped_arange((3,4), xp)
```

negin513 replied:
Done and resolved! Thanks!

dcherian commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -29,6 +29,38 @@ def test_rolling_window(self, xp):
+        arr = testing.shaped_arange((3,4))
+        window_shape = 3
+        axis = 0
+        arr_view = sliding_window_view_cp(arr, window_shape,axis)
```
```suggestion
        arr_view = stride_tricks.sliding_window_view(arr, window_shape,axis)
```

asi1024 replied:
```suggestion
        arr_view = xp.lib.stride_tricks.sliding_window_view(arr, window_shape, axis)
```

dcherian commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +37,109 @@ def as_strided(x, shape=None, strides=None):
+    positions.
+
+
+	Parameters
```
```suggestion
    Parameters
```

dcherian commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +37,109 @@ def as_strided(x, shape=None, strides=None):
+    --------
+    numpy.lib.stride_tricks.as_strided
+
+	Notes
+	--------
+	This function is adapted from numpy.lib.stride_tricks.as_strided.
+
+	Examples
```
```suggestion
    Notes
    --------
    This function is adapted from numpy.lib.stride_tricks.as_strided.

    Examples
```

asi1024 commented:
@negin513 Please confirm if the unit test passes, referring to the following documentation: https://docs.cupy.dev/en/stable/contribution.html#unit-testing

negin513 commented:
@asi1024 I have updated all the tests to use the ` @testing.numpy_cupy_array_equal()` but I am not sure the mechanism underneath. 

> Could you use `testing.numpy_cupy_array_equal()` to compare the outputs of `sliding_window_view` between NumPy and CuPy?

I think I implemented this correctly at this point but I am getting an error with these tests that complains about the datatypes for the assert are different (np vs. cupy) and I am not sure how to approach it. 

Here is the quick snapshot of the failed message from the first test:
```
=================================== FAILURES ===================================
________________________ TestSlidingWindowView.test_1d _________________________
conda-envs/gpu-xdev/lib/python3.9/site-packages/cupy/testing/_loops.py:295: in test_func
    cupy_numpy_result_ndarrays = [
conda-envs/gpu-xdev/lib/python3.9/site-packages/cupy/testing/_loops.py:296: in <listcomp>
    _convert_output_to_ndarray(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

c_out = array([[1., 2.],
       [2., 3.],
       [3., 4.],
       [4., 5.],
       [5., 6.],
       [6., 7.],
       [7., 8.],
       [8., 9.]], dtype=float32)
n_out = array([[1., 2.],
       [2., 3.],
       [3., 4.],
       [4., 5.],
       [5., 6.],
       [6., 7.],
       [7., 8.],
       [8., 9.]], dtype=float32)
sp_name = None, check_sparse_format = True

    def _convert_output_to_ndarray(c_out, n_out, sp_name, check_sparse_format):
        """Checks type of cupy/numpy results and returns cupy/numpy ndarrays.
    
        Args:
            c_out (cupy.ndarray, cupyx.scipy.sparse matrix, cupy.poly1d or scalar):
                cupy result
            n_out (numpy.ndarray, scipy.sparse matrix, numpy.poly1d or scalar):
                numpy result
            sp_name(str or None): Argument name whose value is either
                ``scipy.sparse`` or ``cupyx.scipy.sparse`` module. If ``None``, no
                argument is given for the modules.
            check_sparse_format (bool): If ``True``, consistency of format of
                sparse matrix is also checked. Default is ``True``.
    
        Returns:
            The tuple of cupy.ndarray and numpy.ndarray.
        """
        if sp_name is not None and cupyx.scipy.sparse.issparse(c_out):
            # Sparse output case.
            import scipy.sparse
            assert scipy.sparse.issparse(n_out)
            if check_sparse_format:
                assert c_out.format == n_out.format
            return c_out.A, n_out.A
        if (isinstance(c_out, cupy.ndarray)
                and isinstance(n_out, (numpy.ndarray, numpy.generic))):
            # ndarray output case.
            return c_out, n_out
        if isinstance(c_out, cupy.poly1d) and isinstance(n_out, numpy.poly1d):
            # poly1d output case.
            assert c_out.variable == n_out.variable
            return c_out.coeffs, n_out.coeffs
        if isinstance(c_out, numpy.generic) and isinstance(n_out, numpy.generic):
            # numpy scalar output case.
            return c_out, n_out
        if numpy.isscalar(c_out) and numpy.isscalar(n_out):
            # python scalar output case.
            return cupy.array(c_out), numpy.array(n_out)
>       raise AssertionError(
            'numpy and cupy returns different type of return value:\n'
            'cupy: {}\nnumpy: {}'.format(
                type(c_out), type(n_out)))
E       AssertionError: numpy and cupy returns different type of return value:
E       cupy: <class 'cupy.ndarray'>
E       numpy: <class 'cupy.ndarray'>
```

negin513 commented:
The unit-tests are all passing now! 🚀

asi1024 commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +38,110 @@ def as_strided(x, shape=None, strides=None):
+                             f'got {len(window_shape)} window_shape elements '
+                             f'and `x.ndim` is {x.ndim}.')
+    else:
+        axis = np.core.numeric.normalize_axis_tuple(axis, x.ndim)
```
`np.core.numeric.normalize_axis_tuple` is not NumPy's public API. Please use `cupy._core.internal._normalize_axis_indices` instead.

negin513 replied:
Yes, we initially implemented `cupy._core.internal._normalize_axis_indices` instead of the numpy version. However, the behaviors of these two are slightly different. Unfortunately `cupy._core.internal._normalize_axis_indices` do not accept axis as a list while the numpy equivalent handle the list and this result in an error when using cupy-xarray. 
Please see this discussion in this PR : https://github.com/cupy/cupy/pull/7575#discussion_r1204707652

Here is the error when using the cupy `_nomralize_axis_indices` : 
```
File cupy/core/internal.pyx:403, in cupy.core.internal._normalize_axis_indices()

File cupy/core/internal.pyx:424, in cupy.core.internal._normalize_axis_indices()

TypeError: 'list' object cannot be interpreted as an integer
```

asi1024 replied:
`axis` should be a tuple of int, not a list of int.
https://github.com/cupy/cupy/blob/25c485071ab0a33cdc223c793aa049ab4d84c097/cupy/_core/internal.pyx#L447-L449

negin513 replied:
I have updated xarray to pass axis as a tuple not a list. [Xarray/PR#7983](https://github.com/pydata/xarray/pull/7938)

I also made the corresponding changes here to use `cupy._core.internal._normalize_axis_indices` as you asked. 
This is fully addressed now.

asi1024 commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +38,110 @@ def as_strided(x, shape=None, strides=None):
+                    if np.iterable(window_shape)
+                    else (window_shape,))
+
+    # first convert input to array, possibly keeping subclass
```
Could you add tests for array_like inputs other than cupy.ndarray?

negin513 replied:
Test is added now!

asi1024 commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +38,110 @@ def as_strided(x, shape=None, strides=None):
+    if axis is None:
+        axis = tuple(range(x.ndim))
+        if len(window_shape) != len(axis):
+            raise ValueError(f'Since axis is `None`, must provide '
```
Could you add a test for this ValueError?

negin513 replied:
Thanks for your comment. Test is added now for this.

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -25,10 +25,45 @@ def test_as_strided(self):
         return a_rolling
 
 
+class TestSlidingWindowView(unittest.TestCase):
```
Could you add a test for 0-dim array?

negin513 replied:
Thanks @asi1024 for your comment. Can you please clarify this further? 

I am not sure what you mean by 0-d array test?
Do you mean a scalar basically? If that is the case, the `sliding_window_view` should not work and should spit out an error?

I implemented these tests based on numpy tests for similar functionality and I could not find a 0-d array example there: 
(https://github.com/numpy/numpy/blob/8627dc96202899c01a75f46e3899e2f1118a293f/numpy/lib/tests/test_stride_tricks.py#L400-L430)

negin513 replied:
Just for reference , I can wrote this : 

```
    @testing.numpy_cupy_array_equal()
    def test_0d(self, xp):
        # Create a 0-D array (scalar) for testing
        arr = xp.array(42)
        # Sliding window with window size 1
        window_size = 1 
        arr_view = xp.lib.stride_tricks.sliding_window_view(
            arr, window_size, 0)
        return arr_view    
      
```

Which raises an error for both numpy and cupy : AssertionError: Both cupy and numpy raise exceptions

So I think maybe you mean writing a test to make sure this fails with correct message?

negin513 replied:
I added a test for this that checks the if `ValueError` is raised for 0-d array! Please let me know if this addressed your comment.

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -25,10 +25,45 @@ def test_as_strided(self):
+    def test_1d(self, xp):
+        arr = testing.shaped_arange((3, 4), xp)
+        window_size = 2
+        arr_view = xp.lib.stride_tricks.sliding_window_view(
+            arr, window_size, 0)
```
Could you also check the compatibility of `arr_view.strides`?

negin513 replied:
Hello @asi1024 , Can you please clarify what you mean with "compatibility with `arr_view.stride`"? The `arr_view.strides` work fine when I try to access it but I was not sure what you meant.

Please feel free to push anything you think is appropriate for this.

negin513 commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +38,110 @@ def as_strided(x, shape=None, strides=None):
+
+    window_shape_array = _cupy.array(window_shape)
+    if _cupy.any(window_shape_array < 0):
+        raise ValueError('`window_shape` cannot contain negative values')
```
Thanks! Suggestion applided.

negin513 replied:
Implementing this change caused the tests to fail, but I know what is the culprit here. 

------------
More importantly, I have a question regarding this. I think `_cupy.any` might be more efficient than looping and comparing the results. Although the difference should be very negligible. 

I did some tests on array sizes from 1-20. and except for array size 1 all others were showing better performance using `.any` compared to the loop. 

```
import time
import cupy as cp

for size in range(1, 21):  # Iterate over sizes from 1 to 20
    # Create a 1D array using cp.random.randint()
    array_1d = cp.random.randint(10, size=size)
    
    # Timing the first code snippet
    start_time = time.time()
    if cp.any(array_1d <= 0):
        pass
    end_time = time.time()
    execution_time_1 = (end_time - start_time) * 1e6
    
    # Timing the second code snippet
    start_time = time.time()
    for dim in array_1d:
        if dim <= 0:
            break
    end_time = time.time()
    execution_time_2 = (end_time - start_time) * 1e6

    # Compare the execution times
    faster_method = "_cupy.any()" if execution_time_1 < execution_time_2 else "loop"
    print(f'Summary: Array size: {array_1d.size}, Faster method: {faster_method}, _cupy.any() time: {execution_time_1:.2f} microseconds, Loop time: {execution_time_2:.2f} microseconds')


```
Output: 
```
Summary: Array size: 1, Faster method: loop, _cupy.any() time: 130.65 microseconds, Loop time: 96.08 microseconds
Summary: Array size: 2, Faster method: _cupy.any(), _cupy.any() time: 96.32 microseconds, Loop time: 118.73 microseconds
Summary: Array size: 3, Faster method: _cupy.any(), _cupy.any() time: 84.88 microseconds, Loop time: 180.96 microseconds
Summary: Array size: 4, Faster method: _cupy.any(), _cupy.any() time: 83.92 microseconds, Loop time: 214.34 microseconds
Summary: Array size: 5, Faster method: _cupy.any(), _cupy.any() time: 82.97 microseconds, Loop time: 276.33 microseconds
Summary: Array size: 6, Faster method: _cupy.any(), _cupy.any() time: 82.73 microseconds, Loop time: 113.49 microseconds
Summary: Array size: 7, Faster method: _cupy.any(), _cupy.any() time: 82.73 microseconds, Loop time: 382.66 microseconds
Summary: Array size: 8, Faster method: _cupy.any(), _cupy.any() time: 100.85 microseconds, Loop time: 115.39 microseconds
Summary: Array size: 9, Faster method: _cupy.any(), _cupy.any() time: 85.35 microseconds, Loop time: 269.41 microseconds
Summary: Array size: 10, Faster method: _cupy.any(), _cupy.any() time: 84.64 microseconds, Loop time: 217.44 microseconds
Summary: Array size: 11, Faster method: _cupy.any(), _cupy.any() time: 85.12 microseconds, Loop time: 572.20 microseconds
Summary: Array size: 12, Faster method: _cupy.any(), _cupy.any() time: 85.83 microseconds, Loop time: 267.27 microseconds
Summary: Array size: 13, Faster method: _cupy.any(), _cupy.any() time: 84.40 microseconds, Loop time: 1486.06 microseconds
Summary: Array size: 14, Faster method: _cupy.any(), _cupy.any() time: 89.65 microseconds, Loop time: 322.58 microseconds
Summary: Array size: 15, Faster method: _cupy.any(), _cupy.any() time: 86.55 microseconds, Loop time: 369.55 microseconds
Summary: Array size: 16, Faster method: _cupy.any(), _cupy.any() time: 86.55 microseconds, Loop time: 828.03 microseconds
Summary: Array size: 17, Faster method: _cupy.any(), _cupy.any() time: 87.02 microseconds, Loop time: 267.27 microseconds
Summary: Array size: 18, Faster method: _cupy.any(), _cupy.any() time: 85.83 microseconds, Loop time: 925.30 microseconds
Summary: Array size: 19, Faster method: _cupy.any(), _cupy.any() time: 85.59 microseconds, Loop time: 478.51 microseconds
```
 Still for array size of 1, the loop is faster, so I am okay sticking with the loop method. But for array size 2, 3, it seems like the loop is 3x slower on A100.

asi1024 replied:
Device-to-host synchronization occurs when converting the result of `_cupy.any` to a Python's primitive bool value.

negin513 replied:
This is implemented as you asked.

asi1024 commented on cupy/lib/stride_tricks.py:
```diff
@@ -37,3 +38,110 @@ def as_strided(x, shape=None, strides=None):
+
+
+def sliding_window_view(x, window_shape, axis=None, *,
+                        subok=False, writeable=False):
```
Please fix to raise `NotImplementedError` when `writeable=True` is given.

negin513 replied:
Sure! Just added this to raise `NotImplementedError` when `writeable=True` is given.
Also added a test to make sure it is raised. 

All working now

negin513 commented:
Hello @asi1024 : Thanks for your comments. I have addressed all your above comments, except for [this](https://github.com/cupy/cupy/pull/7575#discussion_r1225321931) which I was not sure what you meant. 

Please let me know if you have any additional concerns and feel free to push anything you think is appropriate to  this PR.
I made sure all the added tests I added are passing: ` python -m pytest test_strided_tricks.py` ... 


Special thanks to @dcherian for his help with this.

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -6,29 +6,98 @@
+        arr = testing.shaped_arange((3, 4), xp)
+        window_size = 2
+        arr_view = xp.lib.stride_tricks.sliding_window_view(arr, window_size, 0)
+        return arr_view
```
Could you check the compatibility of `arr_view.strides` between numpy and cupy?
```suggestion
        assert arr_view.strides == (16, 4, 16)
        return arr_view
```

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -6,29 +6,98 @@
+        arr_view = xp.lib.stride_tricks.sliding_window_view(
+            arr, window_shape=window_shape
+        )
+        return arr_view
```
```suggestion
        assert arr_view.strides == (16, 4, 16, 4)
        return arr_view
```

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -6,29 +6,98 @@
+    def test_2d_with_axis(self, xp):
+        arr = testing.shaped_arange((3, 4), xp)
+        window_shape = 3
+        axis = 0
+        arr_view = xp.lib.stride_tricks.sliding_window_view(arr, window_shape, axis)
```
We want to check with `axis=1` as a non-default axis value.
```suggestion
        axis = 1
        arr_view = xp.lib.stride_tricks.sliding_window_view(
            arr, window_shape, axis)
        assert arr_view.strides == (16, 4, 4)
```

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -6,29 +6,98 @@
+        arr = testing.shaped_arange((3, 4), xp)
+        window_shape = (2, 3)
+        axis = (0, 1)
+        arr_view = xp.lib.stride_tricks.sliding_window_view(arr, window_shape, axis)
```
```suggestion
        arr_view = xp.lib.stride_tricks.sliding_window_view(arr, window_shape, axis)
        assert arr_view.strides == (16, 4, 16, 4)
```

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -6,29 +6,98 @@
+        return arr_view
+
+    def test_0d(self):
+        # Create a 0-D array (scalar) for testing
+        arr = cupy.array(42)
+        # Sliding window with window size 1
+        window_size = 1
+
+        # Test if the correct ValueError is raised!
+        with pytest.raises(ValueError, match="axis 0 is out of bounds"):
+            cupy.lib.stride_tricks.sliding_window_view(arr, window_size, 0)
```
Could you check the raised error messages is equal to NumPy's one?
```suggestion
        for xp in (numpy, cupy):
            # Create a 0-D array (scalar) for testing
            arr = xp.array(42)
            # Sliding window with window size 1
            window_size = 1

            # Test if the correct ValueError is raised!
            with pytest.raises(ValueError, match="axis 0 is out of bounds"):
                xp.lib.stride_tricks.sliding_window_view(arr, window_size, 0)
```

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -6,29 +6,98 @@
+            cupy.lib.stride_tricks.sliding_window_view(arr, window_size, 0)
+
+    def test_window_shape_axis_length_mismatch(self):
+        x = cupy.arange(24).reshape((2, 3, 4))
+        window_shape = (2, 2)
+        axis = None
+
+        # Test if ValueError is raised when len(window_shape) != len(axis)
+        with pytest.raises(ValueError, match="Since axis is `None`"):
+            stride_tricks.sliding_window_view(x, window_shape, axis)
```
```suggestion
        for xp in (numpy, cupy):
            x = xp.arange(24).reshape((2, 3, 4))
            window_shape = (2, 2)
            axis = None

            # Test if ValueError is raised when len(window_shape) != len(axis)
            with pytest.raises(ValueError, match="Since axis is `None`"):
                xp.lib.stride_tricks.sliding_window_view(x, window_shape, axis)
```

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -6,29 +6,98 @@
+        with pytest.raises(ValueError, match="Since axis is `None`"):
+            stride_tricks.sliding_window_view(x, window_shape, axis)
+
+    def test_otherarrays(self):
+        x = numpy.arange(5)
+        arr_view = stride_tricks.sliding_window_view(x, 2)
+        expected = [[0, 1], [1, 2], [2, 3], [3, 4]]
+        testing.assert_array_equal(arr_view, expected)
```
```suggestion
    @testing.numpy_cupy_array_equal()
    def test_arraylike_input(self, xp):
        x = [0, 1, 2, 3, 4]
        arr_view = xp.lib.stride_tricks.sliding_window_view(x, 2)
        assert arr_view.strides == (8, 8)
        return arr_view
```

asi1024 commented:
@negin513 Could you fix for `autopep8`?

negin513 commented:
Thanks for your comments @asi1024 . All comments are now implemented + auto-pep8. 
Is there any other thing that you would like me to address to get this PR merged? Thanks again! --Negin

asi1024 commented:
/test mini

asi1024 commented on tests/cupy_tests/lib_tests/test_strided_tricks.py:
```diff
@@ -25,10 +27,89 @@ def test_as_strided(self):
+
+    @testing.numpy_cupy_array_equal()
+    def test_arraylike_input(self, xp):
+        x = [0, 1, 2, 3, 4]
```
The default dtype for integer seems to be different on Linux and Windows. Could you please fix to use float array?
```suggestion
        x = [0., 1., 2., 3., 4.]
```

asi1024 commented:
/test mini

asi1024 commented:
CI failure seems unrelated to this PR.

asi1024 commented:
@negin513 LGTM! Thanks for your PR!

mdhaber commented:
It looks like this was intended to be public, and it is importable, but the documentation doesn't seem to be rendering.
